File: Assets/Bundles/Client/Resources/TS/Controllers/Global/Damage/Projectile/ProjectileController.lua
-- Compiled with unity-ts v2.1.0-75
local Reflect = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Reflect
local Controller = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Controller
local Object = require("Shared/rbxts_include/node_modules/@easy-games/unity-object-utils/init")
local ClientSignals = require("Client/TS/ClientSignals").ClientSignals
local Entity = require("Shared/TS/Entity/Entity").Entity
local _ItemDefinitions = require("Shared/TS/Item/ItemDefinitions")
local GetItemMeta = _ItemDefinitions.GetItemMeta
local GetItemTypeFromItemId = _ItemDefinitions.GetItemTypeFromItemId
local ItemType = require("Shared/TS/Item/ItemType").ItemType
local Projectile = require("Shared/TS/Projectile/Projectile").Projectile
local LayerUtil = require("Shared/TS/Util/LayerUtil").LayerUtil
local ProjectileCollideClientSignal = require("Client/TS/Controllers/Global/Damage/Projectile/ProjectileCollideClientSignal").ProjectileCollideClientSignal
local ProjectileLaunchedClientSignal = require("Client/TS/Controllers/Global/Damage/Projectile/ProjectileLaunchedClientSignal").ProjectileLaunchedClientSignal
local ProjectileController
do
	ProjectileController = setmetatable({}, {
		__tostring = function()
			return "ProjectileController"
		end,
	})
	ProjectileController.__index = ProjectileController
	function ProjectileController.new(...)
		local self = setmetatable({}, ProjectileController)
		return self:constructor(...) or self
	end
	function ProjectileController:constructor()
		self.prefabInfoByItemType = {}
		for _, itemTypeStr in Object.keys(ItemType) do
			local itemType = itemTypeStr
			local itemMeta = GetItemMeta(itemType)
			if itemMeta.Ammo then
				local projPrefab = AssetBridge:LoadAssetIfExists("Shared/Resources/Prefabs/Projectiles/Ammo/" .. (itemType .. ".prefab"))
				if not projPrefab then
					print("Unable to find asset for ammoItemType: " .. itemType)
					continue
				end
				local rigidbody = projPrefab:GetComponent("Rigidbody")
				local _prefabInfoByItemType = self.prefabInfoByItemType
				local _arg1 = {
					gameObject = projPrefab,
					rigidbody = rigidbody,
				}
				_prefabInfoByItemType[itemType] = _arg1
			end
		end
	end
	function ProjectileController:OnStart()
		ProjectileManager.Instance:onProjectileLaunched(function(easyProjectile, shooterGO)
			local shooterEntity = Entity:FindByGameObject(shooterGO)
			local itemType = GetItemTypeFromItemId(easyProjectile.itemTypeId)
			if not itemType then
				Debug:LogError("Failed to find itemType with id " .. tostring(easyProjectile.itemTypeId))
				return nil
			end
			local projectile = Projectile.new(easyProjectile, itemType, shooterEntity)
			ClientSignals.ProjectileLaunched:Fire(ProjectileLaunchedClientSignal.new(projectile))
		end)
	end
	function ProjectileController:HandleCollision(projectile, collider, hitPoint, normal, velocity)
		local ammoMeta = GetItemMeta(projectile.itemType).Ammo
		local hitEntity = Entity:FindByCollider(collider)
		-- Check if it should be colliding with us.
		if not LayerUtil:LayerIsInMask(collider.gameObject.layer, ammoMeta.projectileHitLayerMask) then
			return false
		end
		local projectileHitSignal = ProjectileCollideClientSignal.new(projectile, hitPoint, normal, velocity, hitEntity)
		ClientSignals.ProjectileCollide:Fire(projectileHitSignal)
		return true
	end
end
-- (Flamework) ProjectileController metadata
Reflect.defineMetadata(ProjectileController, "identifier", "Bundles/Client/Controllers/Global/Damage/Projectile/ProjectileController@ProjectileController")
Reflect.defineMetadata(ProjectileController, "flamework:implements", { "$:flamework@OnStart" })
Reflect.decorate(ProjectileController, "$:flamework@Controller", Controller, { {} })
return {
	ProjectileController = ProjectileController,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Typescript~/src/Client/Controllers/Global/Damage/Projectile/ProjectileController.ts
import { Controller, OnStart } from "@easy-games/flamework-core";
import Object from "@easy-games/unity-object-utils";
import { ClientSignals } from "Client/ClientSignals";
import { Entity } from "Shared/Entity/Entity";
import { ItemType } from "Shared/Item/ItemType";
import { Projectile } from "Shared/Projectile/Projectile";
import { LayerUtil } from "Shared/Util/LayerUtil";
import { ProjectileCollideClientSignal } from "./ProjectileCollideClientSignal";
import { ProjectileLaunchedClientSignal } from "./ProjectileLaunchedClientSignal";
import { ItemUtil } from "../../../../../Shared/Item/ItemUtil";

@Controller({})
export class ProjectileController implements OnStart {
	private prefabInfoByItemType = new Map<ItemType, { gameObject: GameObject; rigidbody: Rigidbody }>();

	constructor() {
		for (const itemTypeStr of Object.keys(ItemType)) {
			const itemType = itemTypeStr as ItemType;
			const itemMeta = ItemUtil.GetItemMeta(itemType);

			if (itemMeta.Ammo) {
				const projPrefab = AssetBridge.LoadAssetIfExists(
					`Shared/Resources/Prefabs/Projectiles/Ammo/${itemType}.prefab`,
				) as GameObject;

				if (!projPrefab) {
					print(`Unable to find asset for ammoItemType: ${itemType}`);
					continue;
				}

				const rigidbody = projPrefab.GetComponent<Rigidbody>();

				this.prefabInfoByItemType.set(itemType, { gameObject: projPrefab, rigidbody: rigidbody });
			}
		}
	}

	public OnStart(): void {
		ProjectileManager.Instance.onProjectileLaunched((easyProjectile, shooterGO) => {
			const shooterEntity = Entity.FindByGameObject(shooterGO);
			const itemType = ItemUtil.GetItemTypeFromItemId(easyProjectile.itemTypeId);
			if (!itemType) {
				Debug.LogError("Failed to find itemType with id " + easyProjectile.itemTypeId);
				return;
			}
			const projectile = new Projectile(easyProjectile, itemType, shooterEntity);
			ClientSignals.ProjectileLaunched.Fire(new ProjectileLaunchedClientSignal(projectile));
		});
	}

	public HandleCollision(
		projectile: Projectile,
		collider: Collider,
		hitPoint: Vector3,
		normal: Vector3,
		velocity: Vector3,
	): boolean {
		const ammoMeta = ItemUtil.GetItemMeta(projectile.itemType).Ammo!;
		const hitEntity = Entity.FindByCollider(collider);

		// Check if it should be colliding with us.
		if (!LayerUtil.LayerIsInMask(collider.gameObject.layer, ammoMeta.projectileHitLayerMask)) {
			return false;
		}

		const projectileHitSignal = new ProjectileCollideClientSignal(
			projectile,
			hitPoint,
			normal,
			velocity,
			hitEntity,
		);
		ClientSignals.ProjectileCollide.Fire(projectileHitSignal);

		return true;
	}
}
